### PR TITLE
Add sniff to enforce Copyright information in every file

### PR DESCRIPTION
**Pull request by amazd**

### CODE REVIEW

File: tests/Behance/Sniffs/Comments/EnforceCopyrightCommentSniffTest.inc.fixed [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+* and its suppliers and are protected by all applicable intellectual
+* property laws, including trade secret and copyright laws.
+* Dissemination of this information or reproduction of this material
+* is strictly forbidden unless prior written permission is obtained
+* from Adobe.
+**************************************************************************/
+
+function foo() {
+    echo 'hello world';
+}
```
Comment (andyroo2000) [inline tests/Behance/Sniffs/Comments/EnforceCopyrightCommentSniffTest.inc.fixed]: Where did `foo2` go?

File: Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+   * @param PHP_CodeSniffer_File $phpcsFile The file where the token was found.
+   * @param int                  $stackPtr  The position in the stack where
+   *                                        the token was found.
+   *
+   * @return void
+   */
+  public function process(PHP_CodeSniffer_File $phpcsFile, $stackPtr) {
+
+    /***
+     * We are looking for a comment block at the top of php files in the following format:
```
Comment (mikesherov) [inline Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php]: No need for this comment. The test will describe this just fine!

File: Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+     // just look at the first 5 lines of a file
+    if (!preg_match("/\* ADOBE CONFIDENTIAL/", $phpcsFile->getTokensAsString(0, 5))) {
+
+      $error = 'No Adobe Copyright block found at the top of the file!';
+      $phpcsFile->addError($error, $stackPtr);
+
+    }
+
+    return (count($phpcsFile->getTokens()) + 1); // skips checks on the rest of the file
```
Comment (mikesherov) [inline Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php]: I don’t understand what this is doing or how.

File: Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+     * * ADOBE CONFIDENTIAL
+     * * ___________________
+     * *
+     * * Copyright 2018 Adobe
+     * * All Rights Reserved.
+     * *
+     */
+
+     // just look at the first 5 lines of a file
+    if (!preg_match("/\* ADOBE CONFIDENTIAL/", $phpcsFile->getTokensAsString(0, 5))) {
```
Comment (mikesherov) [inline Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php]: I think it should auto fix, so it should know the exact text to place here, including the year.

File: Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+     * /*************************************************************************
+     * * ADOBE CONFIDENTIAL
+     * * ___________________
+     * *
+     * * Copyright 2018 Adobe
+     * * All Rights Reserved.
+     * *
+     */
+
+     // just look at the first 5 lines of a file
```
Comment (mikesherov) [inline Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php]: Unnecessary comment. Always comment *why*, not what. If you need a comment for what code is doing, refactor to explaining function name.

File: Makefile [hunk 6]
```diff
@@ -10,7 +10,7 @@ ci:
 	./vendor/bin/parallel-lint -j 10 $(SPACE_SEPARATED_CHANGED_PHP_FILES)
 
 	# Static Analysis
-	./vendor/bin/phpcs --standard=./Behance -s -n $(SPACE_SEPARATED_CHANGED_PHP_FILES)
+	./vendor/bin/phpcs --exclude=Behance.Comments.EnforceCopyrightComment --standard=./Behance -s -n $(SPACE_SEPARATED_CHANGED_PHP_FILES)
```
Comment (mikesherov) [inline Makefile]: We need to enforce this.

File: Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php (hunk 5 above)
Comment (amazd) [reply Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php]: I wasn't sure if it was clear enough that the call to ->getTokensAsString(0, 5) would return the first 5 lines of a file. If you feels it does, I can remove.

File: Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php (hunk 3 above)
Comment (amazd) [reply Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php]: @mikesherov it's based on this comment:
https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Sniffs/Sniff.php#L74

File: Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php (hunk 5 above)
Comment (mikesherov) [reply Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php]: Yeah, remove. But also, if we’re going to autofix, this code will change anyway :-)

File: Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+   */
+  public function process(PHP_CodeSniffer_File $phpcsFile, $stackPtr) {
+
+    $this->_phpcsFile = $phpcsFile;
+    $this->_tokens = $this->_phpcsFile->getTokens();
+
+    $exclude = true;
+    $next = $this->_phpcsFile->findNext(T_WHITESPACE, ($stackPtr + 1), null, $exclude);
+
+    if ($this->_tokens[$next]['type'] != 'T_COMMENT'
```
Comment (andyroo2000) [inline Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php]: may as well use `!==` here

File: Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+      $error = 'No Adobe Copyright block found at the top of the file!';
+
+      $should_fix = $this->_phpcsFile->addFixableError($error, $next, 'CloseBraceNewLine');
+
+      if ($should_fix) {
+        $this->_phpcsFile->fixer->addContentBefore($stackPtr + 1, $this->_phpcsFile->eolChar . sprintf($this->_copyright_block, date('Y')) . $this->_phpcsFile->eolChar);
+      }
+    }
+
+    $nextArrow = $next;
```
Comment (mikesherov) [inline Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php]: rename to `$next_comment`

File: Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    $block_num_of_lines = mb_substr_count($this->_copyright_block, "\n");
+
+    while (($nextArrow = $this->_phpcsFile->findNext(T_COMMENT, ($nextArrow + 1), null)) !== false) {
+      if (preg_match($this->_block_signature, $this->_tokens[$nextArrow + 1]['content'])
+       && preg_match($this->_end_block_signature, $this->_tokens[$nextArrow + $block_num_of_lines]['content'])) {
+        $error = 'Duplicate Copyright block found!';
+        $should_fix = $this->_phpcsFile->addFixableError($error, $nextArrow, 'CloseBraceNewLine');
+        if ($should_fix) {
+          $this->_removeDuplicateBlock($nextArrow);
```
Comment (mikesherov) [inline Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php]: I don't understand why this is a loop? you're checking each comment... for each duplicate found, remove it.

File: Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+   *
+   * @param PHP_CodeSniffer_File $phpcsFile The file where the token was found.
+   * @param int                  $stackPtr  The position in the stack where
+   *                                        the token was found.
+   *
+   * @return void
+   */
+  public function process(PHP_CodeSniffer_File $phpcsFile, $stackPtr) {
+
+    $this->_phpcsFile = $phpcsFile;
```
Comment (mikesherov) [inline Behance/Sniffs/Comments/EnforceCopyrightCommentSniff.php]: no need for member variables because you're going to remove that private method.

**Pull request merged**